### PR TITLE
Add critical fours variants

### LINKED ISSUE #2168: Variant (Re)Proposal: Critical Fours
Fairly self explanatory, only one 4 of each suit.

This is nice as an alternative to dark suits for variants that impose higher efficiency requirements, since black suits come with a significant amount of luck around bottom-decks that is out of your control. Crit 4s gives a similar efficiency bar but the bottom-deck with normal rules still lets you get N-1 points and with bottom deck blind plays it is very unlikely to prevent a perfect on its own.

I have implemented this on a fork and we've played a game of it. Nothing too special IMO but the players mostly agreed that they prefer it to dark suits. https://youtu.be/Rs22PPtL3lI

### PR DESCRIPTION
**Pull request by bhamrick**
Closes #2168. Tested a no variant game and a critical fours game to completion and they both appeared to work as expected.

### CODE REVIEW

**Comment by bhamrick** [comment]
Hang on, changes don't seem to include the fix sorry.

**Comment by bhamrick** [comment]
Ok, the changes _do_ include the fix. I was confused a minute ago because the implementation in client/src/game/rules/deck.ts doesn't match the implementation in server/src/game_deck.go (the ts version sets amountToAdd to 2 at the top of the function, while the go version leaves it at the default value of 0).

**Comment by Zamiell** [comment]
these functions suck, can you refactor them so that they return values instead of use a variable that is mutated?

**Comment by Zamiell** [comment]
that will prevent future bugs

**Comment by bhamrick** [comment]
Sure.

**Comment by bhamrick** [comment]
I quickly tested what I think is all the game types:

* Normal
* Reversed
* Up or Down
* One of Each (single and double dark)
* Crit Fours

and the decks had the expected distributions.

File: client/src/game/rules/deck.ts [hunk 1]
```diff
@@ -30,33 +33,41 @@ export function numCopiesOfCard(
   rank: number,
   variant: Variant,
 ): number {
-  // In a normal suit of Hanabi,
-  // there are three 1's, two 2's, two 3's, two 4's, and one 5
-  let amountToAdd = 2;
+  // This implementation mirrors numCopiesOfCard in server/src/game_deck.go
```
Comment (Zamiell) [inline client/src/game/rules/deck.ts:36]: good thank you

**Pull request merged**